nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,335 @@
+    params_->v_linear_min, params_->v_linear_max, params_->v_angular_max);
+
+  // Initialize footprint collision checker
+  collision_checker_ = std::make_unique<nav2_costmap_2d::
+      FootprintCollisionChecker<nav2_costmap_2d::Costmap2D *>>(costmap_ros_->getCostmap());
+
+  // Publishers
+  transformed_plan_pub_ = node->create_publisher<nav_msgs::msg::Path>("transformed_global_plan", 1);
+  local_plan_pub_ = node->create_publisher<nav_msgs::msg::Path>("local_plan", 1);
+  motion_target_pub_ = node->create_publisher<geometry_msgs::msg::PointStamped>("motion_target", 1);
+  slowdown_pub_ = node->create_publisher<visualization_msgs::msg::Marker>("slowdown", 1);
+
+  RCLCPP_INFO(logger_, "Configured Graceful Motion Controller: %s", plugin_name_.c_str());
+}
+
+void GracefulMotionController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_.reset();
+  local_plan_pub_.reset();
+  motion_target_pub_.reset();
+  slowdown_pub_.reset();
+}
+
+void GracefulMotionController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_->on_activate();
+  local_plan_pub_->on_activate();
+  motion_target_pub_->on_activate();
+  slowdown_pub_->on_activate();
+}
+
+void GracefulMotionController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_->on_deactivate();
+  local_plan_pub_->on_deactivate();
+  motion_target_pub_->on_deactivate();
+  slowdown_pub_->on_deactivate();
+}
+
+geometry_msgs::msg::TwistStamped GracefulMotionController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & /*velocity*/,
+  nav2_core::GoalChecker * goal_checker)
+{
+  std::lock_guard<std::mutex> param_lock(param_handler_->getMutex());
+
+  // Update for the current goal checker's state
+  geometry_msgs::msg::Pose pose_tolerance;
+  geometry_msgs::msg::Twist velocity_tolerance;
+  if (!goal_checker->getTolerances(pose_tolerance, velocity_tolerance)) {
+    RCLCPP_WARN(logger_, "Unable to retrieve goal checker's tolerances!");
+  } else {
+    goal_dist_tolerance_ = pose_tolerance.position.x;
+  }
+
+  // Transform path to robot base frame and publish it
+  auto transformed_plan = path_handler_->transformGlobalPlan(
+    pose, params_->max_robot_pose_search_dist);
+  transformed_plan_pub_->publish(transformed_plan);
+
+  // Get the particular point on the path at the motion target distance and publish it
+  auto motion_target = getMotionTarget(params_->motion_target_dist, transformed_plan);
+  auto motion_target_point = nav2_graceful_motion_controller::createMotionTargetMsg(motion_target);
```

Comment:
I think you will find this performs splendidly poorly in the real world.

 * The graceful law doesn't necessarily follow the path exactly
 * If you only pick the farthest out motion point once, then if the graceful law takes you off the path, and you later fail in the collision checking, you get no velocity command and end up needing to globally replan

What my graceful controller does is an iterative approach:

 * Select a candidate motion target (it must satisfy the rules, such as not too far away from robot)
 * Run the control law
 * Forward simulate with collisions
 * If there is a collision, repeat by selecting a candidate motion target that is a bit closer on the path (until you hit a minimum distance threshold beyond which we have to return to the global planner)

Reply:
I agree. I tested in our robot and, unless you set a motion point very close to the robot, it will not follow the path exactly.